### PR TITLE
Add new offer status watcher facade

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Add a new facade to support watching the status of offers. As is usual with this type of change, there's a lot of boilerplate. Subsequent work will make use of this new functionality.

## QA steps

Run the tests - there's nothing to test yet until the next branch.

### CODE REVIEW

File: apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go [hunk 1]
```diff
@@ -403,6 +408,68 @@ func (api *CrossModelRelationsAPI) WatchRelationsSuspendedStatus(
...
+		}
+		_, ok := <-w.Changes()
+		if !ok {
+			results.Results[i].Error = common.ServerError(watcher.EnsureErr(w))
+			continue
+		}
+		results.Results[i].OfferStatusWatcherId = api.resources.Register(w)
+		change, err := commoncrossmodel.GetOfferStatusChange(api.st, arg.OfferUUID)
+		if err != nil {
+			results.Results[i].Error = common.ServerError(err)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go]: done

File: apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go [hunk 2]
```diff
@@ -403,6 +408,68 @@ func (api *CrossModelRelationsAPI) WatchRelationsSuspendedStatus(
...
+			continue
+		}
+		_, ok := <-w.Changes()
+		if !ok {
+			results.Results[i].Error = common.ServerError(watcher.EnsureErr(w))
+			continue
+		}
+		results.Results[i].OfferStatusWatcherId = api.resources.Register(w)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go]: yup

File: api/watcher/watcher_test.go [hunk 3]
```diff
@@ -394,6 +395,106 @@ func (s *watcherSuite) TestRelationStatusWatcherDeadRelation(c *gc.C) {
...
+
+	// Start watching for a relation change.
+	client := crossmodelrelations.NewClient(s.stateAPI)
+	w, err := client.WatchOfferStatus(params.OfferArg{
+		OfferUUID: offer.OfferUUID,
+		Macaroons: macaroon.Slice{mac},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	stop := func() {
+		c.Assert(worker.Stop(w), jc.ErrorIsNil)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/watcher/watcher_test.go]: ok

File: api/watcher/watcher.go [hunk 4]
```diff
@@ -383,6 +384,78 @@ func (w *relationStatusWatcher) Changes() watcher.RelationStatusChannel {
...
+					Data:    ch.Status.Data,
+					Since:   ch.Status.Since,
+				},
+			}
+		}
+		return result
+	}
+	changes := copyChanges(initialChanges)
+	for {
+		select {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/watcher/watcher.go]: yeah, good point

**Comment by axw** [comment]
Thanks for adding the coalescence test!

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**